pkg/router/routeapihelpers/validation.go

```diff
@@ -353,6 +363,34 @@ func validateInsecureEdgeTerminationPolicy(tls *routev1.TLSConfig, fldPath *fiel
 	return nil
 }
 
+// isSelfSignedCert determines if a provided certificate is
+// self-signed by checking if the issuer is equal to the subject.
+func isSelfSignedCert(cert *x509.Certificate) bool {
+	return bytes.Equal(cert.RawIssuer, cert.RawSubject)
```

Comment:
There's a discussion of how to tell when a cert is self-signed in https://security.stackexchange.com/questions/93162/how-to-know-if-certificate-is-self-signed, and one example shows "Just because the fields have the same value that does not mean the certificate is self-signed".  Do you think that could apply here?

Reply:
Okay, I dug into this even more, because I want to be _really_ confident about this check. 

I traced the HAProxy certificate validation back to [ssl_sock_load_cert_chain](https://github.com/haproxy/haproxy/blob/4332fed6c1031fa2f28af4777c45fd84a724ce99/src/ssl_sock.c#L2671) which calls [SSL_CTX_use_certificate](https://github.com/openssl/openssl/blob/b85e6f534906f0bf9114386d227e481d2336a0ff/ssl/ssl_rsa.c#L239) which eventually calls [ssl_security_cert_sig](https://github.com/openssl/openssl/blob/b85e6f534906f0bf9114386d227e481d2336a0ff/ssl/t1_lib.c#L3649), and this function has the logic that skips checks for self-signed certs:
```c
    /* Don't check signature if self signed */
    if ((X509_get_extension_flags(x) & EXFLAG_SS) != 0)
        return 1;
```

So, I don't think I should get too hung up on the "perfect" way to check for self-signed. Really, our goal should be to match what OpenSSL views as self-signed, so we only reject what OpenSSL (and therefore HAProxy) would reject. OpenSSL considers a cert self signed if:
1. Subject == Issuer
2. Authority Key ID == Subject Key ID
3. Signature Algorithm == Public Key Algorithm

3 is a bit awkward to do with the Go Library, but I guess not horrible. Code has been updated to do all 3. I prefer mirroring what OpenSSL is doing as it's a more exact validation. Sorry for the additional churn.